File: DeenTime/backend/DeenTime.Api/Services/IslamicContent/HadithProviderClient.cs
using System.Globalization;
using System.Net;
using System.Text.Json;
using Microsoft.Extensions.Options;

namespace DeenTime.Api.Services.IslamicContent;

public sealed record HadithProviderBook(
    int Id,
    string Slug,
    string Name,
    string WriterName,
    string? AboutWriter,
    string? WriterDeath,
    int HadithCount,
    int ChapterCount);

public sealed record HadithProviderChapter(
    int Id,
    string BookSlug,
    int ChapterNumber,
    string? Arabic,
    string? English,
    string? Urdu);

public sealed record HadithProviderRecord(
    int Id,
    string HadithNumber,
    string BookSlug,
    int? ChapterNumber,
    int? Volume,
    string? Status,
    string? EnglishNarrator,
    string? UrduNarrator,
    string? English,
    string? Urdu,
    string? Arabic,
    string? HeadingEnglish,
    string? HeadingUrdu,
    string? HeadingArabic);

public sealed record HadithProviderPage(
    int CurrentPage,
    int LastPage,
    int PerPage,
    int Total,
    IReadOnlyList<HadithProviderRecord> Records);

public sealed class HadithProviderClient
{
    public const string ProviderName = "hadithapi";
    public static readonly string[] Languages = ["ar", "en", "ur"];

    private readonly HttpClient _http;
    private readonly IslamicContentOptions _options;
    private readonly SemaphoreSlim _requestGate = new(1, 1);
    private DateTimeOffset _nextRequestAtUtc = DateTimeOffset.MinValue;

    public HadithProviderClient(HttpClient http, IOptions<IslamicContentOptions> options)
    {
        _http = http;
        _options = options.Value;
    }

    public bool IsConfigured => !string.IsNullOrWhiteSpace(_options.HadithApiKey);

    public async Task<IReadOnlyList<HadithProviderBook>> GetBooksAsync(CancellationToken cancellationToken = default)
    {
        using var document = await GetDocumentAsync("books", null, cancellationToken);
        var books = FindArray(document.RootElement, "books");
        return books
            .Select(item => new HadithProviderBook(
                GetInt(item, "id") ?? 0,
                GetString(item, "bookSlug", "book_slug", "slug") ?? "",
                GetString(item, "bookName", "book_name", "name") ?? "",
                GetString(item, "writerName", "writer_name") ?? "",
                GetString(item, "aboutWriter", "about_writer"),
                GetString(item, "writerDeath", "writer_death"),
                GetInt(item, "hadiths_count", "hadithCount", "hadith_count") ?? 0,
                GetInt(item, "chapters_count", "chapterCount", "chapter_count") ?? 0))
            .Where(book => book.Id > 0 && !string.IsNullOrWhiteSpace(book.Slug))
            .ToArray();
    }

    public async Task<IReadOnlyList<HadithProviderChapter>> GetChaptersAsync(
        string bookSlug,
        CancellationToken cancellationToken = default)
    {
        EnsureSlug(bookSlug);
        using var document = await GetDocumentAsync($"{bookSlug}/chapters", null, cancellationToken);
        var chapters = FindArray(document.RootElement, "chapters");
        return chapters
            .Select(item => new HadithProviderChapter(
                GetInt(item, "id") ?? 0,
                GetString(item, "bookSlug", "book_slug") ?? bookSlug,
                GetInt(item, "chapterNumber", "chapter_number") ?? 0,
                GetString(item, "chapterArabic", "chapter_arabic"),
                GetString(item, "chapterEnglish", "chapter_english"),
                GetString(item, "chapterUrdu", "chapter_urdu")))
            .Where(chapter => chapter.ChapterNumber > 0)
            .ToArray();
    }

    public async Task<HadithProviderPage> GetHadithPageAsync(
        string bookSlug,
        int page,
        int pageSize = 200,
        CancellationToken cancellationToken = default)
    {
        EnsureSlug(bookSlug);
        if (page < 1) throw new ArgumentOutOfRangeException(nameof(page));
        pageSize = Math.Clamp(pageSize, 1, 200);

        using var document = await GetDocumentAsync(
            "hadiths",
            new Dictionary<string, string?>
            {
                ["book"] = bookSlug,
                ["paginate"] = pageSize.ToString(CultureInfo.InvariantCulture),
                ["page"] = page.ToString(CultureInfo.InvariantCulture)
            },
            cancellationToken);

        if (!TryFindProperty(document.RootElement, "hadiths", out var envelope) || envelope.ValueKind != JsonValueKind.Object)
            throw new IslamicContentProviderException("The Hadith provider returned an unexpected response.");

        var records = FindArray(envelope, "data")
            .Select(item => ParseHadith(item, bookSlug))
            .Where(item => item.Id > 0 && !string.IsNullOrWhiteSpace(item.HadithNumber))
            .ToArray();

        return new HadithProviderPage(
            GetInt(envelope, "current_page", "currentPage") ?? page,
            Math.Max(1, GetInt(envelope, "last_page", "lastPage") ?? page),
            GetInt(envelope, "per_page", "perPage") ?? pageSize,
            GetInt(envelope, "total") ?? records.Length,
            records);
    }

    private async Task<JsonDocument> GetDocumentAsync(
        string relativePath,
        IReadOnlyDictionary<string, string?>? parameters,
        CancellationToken cancellationToken)
    {
        if (!IsConfigured)
            throw new IslamicContentProviderException("The Hadith provider key has not been configured on the server.");

        var query = new List<KeyValuePair<string, string?>>();
        if (parameters is not null) query.AddRange(parameters);
        query.Add(new KeyValuePair<string, string?>("apiKey", _options.HadithApiKey));

        var uri = relativePath.TrimStart('/') + "?" + string.Join("&", query.Select(pair =>
            $"{Uri.EscapeDataString(pair.Key)}={Uri.EscapeDataString(pair.Value ?? string.Empty)}"));

        const int maxAttempts = 8;
        for (var attempt = 1; attempt <= maxAttempts; attempt++)
        {
            await _requestGate.WaitAsync(cancellationToken);
            try
            {
                var throttleDelay = _nextRequestAtUtc - DateTimeOffset.UtcNow;
                if (throttleDelay > TimeSpan.Zero)
                    await Task.Delay(throttleDelay, cancellationToken);

                using var response = await _http.GetAsync(uri, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
                _nextRequestAtUtc = DateTimeOffset.UtcNow.AddMilliseconds(850);

                if (response.StatusCode == HttpStatusCode.TooManyRequests)
                {
                    var retryAfter = response.Headers.RetryAfter?.Delta
                        ?? (response.Headers.RetryAfter?.Date - DateTimeOffset.UtcNow)
                        ?? TimeSpan.FromSeconds(Math.Min(60, 3 * Math.Pow(2, attempt - 1)));
                    if (retryAfter < TimeSpan.FromSeconds(2)) retryAfter = TimeSpan.FromSeconds(2);
                    if (retryAfter > TimeSpan.FromMinutes(2)) retryAfter = TimeSpan.FromMinutes(2);
                    _nextRequestAtUtc = DateTimeOffset.UtcNow.Add(retryAfter);

                    if (attempt == maxAttempts)
                        throw new IslamicContentProviderException(
                            "The Hadith provider rate limit remained active after automatic retries.");
                    continue;
                }

                if (!response.IsSuccessStatusCode)
                    throw new IslamicContentProviderException(
                        $"The Hadith provider returned HTTP {(int)response.StatusCode} ({response.StatusCode}).");

                await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken);
                return await JsonDocument.ParseAsync(stream, cancellationToken: cancellationToken);
            }
            catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
            {
                throw new IslamicContentProviderException("The Hadith provider timed out.");
            }
            catch (JsonException ex)
            {
                throw new IslamicContentProviderException("The Hadith provider returned an invalid response.", ex);
            }
            catch (HttpRequestException ex)
            {
                if (attempt == maxAttempts)
                    throw new IslamicContentProviderException("The Hadith provider is currently unavailable.", ex);
                _nextRequestAtUtc = DateTimeOffset.UtcNow.AddSeconds(Math.Min(30, attempt * 2));
            }
            finally
            {
                _requestGate.Release();
            }
        }

        throw new IslamicContentProviderException("The Hadith provider is currently unavailable.");
    }

    private static HadithProviderRecord ParseHadith(JsonElement item, string fallbackBookSlug)
    {
        int? chapterNumber = GetInt(item, "chapterNumber", "chapter_number");
        if (TryFindProperty(item, "chapter", out var chapter) && chapter.ValueKind == JsonValueKind.Object)
            chapterNumber ??= GetInt(chapter, "chapterNumber", "chapter_number", "id");
        chapterNumber ??= GetInt(item, "chapterId", "chapter_id");

        return new HadithProviderRecord(
            GetInt(item, "id") ?? 0,
            GetString(item, "hadithNumber", "hadith_number") ?? "",
            GetString(item, "bookSlug", "book_slug") ?? fallbackBookSlug,
            chapterNumber,
            GetInt(item, "volume"),
            GetString(item, "status"),
            GetString(item, "englishNarrator", "english_narrator"),
            GetString(item, "urduNarrator", "urdu_narrator"),
            GetString(item, "hadithEnglish", "hadith_english"),
            GetString(item, "hadithUrdu", "hadith_urdu"),
            GetString(item, "hadithArabic", "hadith_arabic"),
            GetString(item, "headingEnglish", "heading_english"),
            GetString(item, "headingUrdu", "heading_urdu"),
            GetString(item, "headingArabic", "heading_arabic"));
    }

    private static IReadOnlyList<JsonElement> FindArray(JsonElement root, string propertyName)
    {
        if (TryFindProperty(root, propertyName, out var direct) && direct.ValueKind == JsonValueKind.Array)
            return direct.EnumerateArray().ToArray();

        if (TryFindProperty(root, "data", out var data) && data.ValueKind == JsonValueKind.Object &&
            TryFindProperty(data, propertyName, out var nested) && nested.ValueKind == JsonValueKind.Array)
            return nested.EnumerateArray().ToArray();

        return [];
    }

    private static string? GetString(JsonElement element, params string[] names)
    {
        foreach (var name in names)
        {
            if (!TryFindProperty(element, name, out var value) || value.ValueKind is JsonValueKind.Null or JsonValueKind.Undefined)
                continue;
            return value.ValueKind == JsonValueKind.String ? value.GetString() : value.ToString();
        }
        return null;
    }

    private static int? GetInt(JsonElement element, params string[] names)
    {
        var value = GetString(element, names);
        if (int.TryParse(value, NumberStyles.Integer, CultureInfo.InvariantCulture, out var number)) return number;
        if (decimal.TryParse(value, NumberStyles.Number, CultureInfo.InvariantCulture, out var decimalNumber))
            return (int)decimalNumber;
        return null;
    }

    private static bool TryFindProperty(JsonElement element, string name, out JsonElement value)
    {
        if (element.ValueKind == JsonValueKind.Object)
        {
            foreach (var property in element.EnumerateObject())
            {
                if (property.Name.Equals(name, StringComparison.OrdinalIgnoreCase))
                {
                    value = property.Value;
                    return true;
                }
            }
        }
        value = default;
        return false;
    }

    private static void EnsureSlug(string slug)
    {
        if (string.IsNullOrWhiteSpace(slug) || slug.Length > 80 ||
            slug.Any(character => !(char.IsAsciiLetterOrDigit(character) || character is '-' or '_')))
            throw new ArgumentException("Invalid Hadith book slug.", nameof(slug));
    }
}

File: DeenTime/backend-node/src/integrations/islamic-content/hadith-provider.client.ts
import { Inject, Injectable, Optional } from '@nestjs/common';
import { ArgumentError, IslamicContentProviderError } from './errors.js';
import { IslamicContentOptions } from './islamic-content-options.js';
import {
  PROVIDER_CLIENT_OPTIONS,
  classifyFetchError,
  delay as realDelay,
  ensureTrailingSlash,
  escapeDataString,
  httpStatusName,
  requestSignal,
  resolveClock,
  resolveFetch,
  type ProviderClientOptions,
} from './provider-http.js';

export interface HadithProviderBook {
  id: number;
  slug: string;
  name: string;
  writerName: string;
  aboutWriter: string | null;
  writerDeath: string | null;
  hadithCount: number;
  chapterCount: number;
}

export interface HadithProviderChapter {
  id: number;
  bookSlug: string;
  chapterNumber: number;
  arabic: string | null;
  english: string | null;
  urdu: string | null;
}

export interface HadithProviderRecord {
  id: number;
  hadithNumber: string;
  bookSlug: string;
  chapterNumber: number | null;
  volume: number | null;
  status: string | null;
  englishNarrator: string | null;
  urduNarrator: string | null;
  english: string | null;
  urdu: string | null;
  arabic: string | null;
  headingEnglish: string | null;
  headingUrdu: string | null;
  headingArabic: string | null;
}

export interface HadithProviderPage {
  currentPage: number;
  lastPage: number;
  perPage: number;
  total: number;
  records: HadithProviderRecord[];
}

const RETRY = Symbol('retry');
type AttemptOutcome = { document: unknown } | typeof RETRY;

const SLUG_PATTERN = /^[A-Za-z0-9_-]+$/;
const INT32_MAX = 2_147_483_647;
const INT32_MIN = -2_147_483_648;

function asObject(value: unknown): Record<string, unknown> | null {
  return value !== null && typeof value === 'object' && !Array.isArray(value) ? (value as Record<string, unknown>) : null;
}

/** TryFindProperty: case-insensitive property lookup; undefined when absent (JSON null counts as absent). */
export function findHadithProperty(element: unknown, name: string): unknown {
  const object = asObject(element);
  if (object === null) return undefined;
  const lowered = name.toLowerCase();
  for (const key of Object.keys(object)) {
    if (key.toLowerCase() === lowered) return object[key] ?? undefined;
  }
  return undefined;
}

/** FindArray: `root[name]` when it is an array, else `root.data[name]`, else []. */
export function findHadithArray(root: unknown, name: string): unknown[] {
  const direct = findHadithProperty(root, name);
  if (Array.isArray(direct)) return direct;
  const data = findHadithProperty(root, 'data');
  if (asObject(data) !== null) {
    const nested = findHadithProperty(data, name);
    if (Array.isArray(nested)) return nested;
  }
  return [];
}

/** GetString: the first present property among `names`, as JsonElement.ToString() would render it. */
export function getHadithString(element: unknown, ...names: string[]): string | null {
  for (const name of names) {
    const value = findHadithProperty(element, name);
    if (value === undefined) continue;
    if (typeof value === 'string') return value;
    if (typeof value === 'number') return String(value);
    if (typeof value === 'boolean') return value ? 'True' : 'False';
    return JSON.stringify(value);
  }
  return null;
}

/** GetInt: int.TryParse, then decimal.TryParse truncated toward zero. */
export function getHadithInt(element: unknown, ...names: string[]): number | null {
  const text = getHadithString(element, ...names);
  if (text === null) return null;
  const trimmed = text.trim();
  if (/^[+-]?\d+$/.test(trimmed)) {
    const integer = Number(trimmed);
    return integer >= INT32_MIN && integer <= INT32_MAX ? integer : null;
  }
  const numeric = trimmed.replace(/,/g, '');
  if (!/^[+-]?(?:\d+\.?\d*|\.\d+)$/.test(numeric)) return null;
  const truncated = Math.trunc(Number(numeric));
  return truncated >= INT32_MIN && truncated <= INT32_MAX ? truncated : null;
}

function parseHadith(item: unknown, fallbackBookSlug: string): HadithProviderRecord {
  let chapterNumber = getHadithInt(item, 'chapterNumber', 'chapter_number');
  const chapter = findHadithProperty(item, 'chapter');
  if (chapterNumber === null && asObject(chapter) !== null) chapterNumber = getHadithInt(chapter, 'chapterNumber', 'chapter_number', 'id');
  chapterNumber ??= getHadithInt(item, 'chapterId', 'chapter_id');

  return {
    id: getHadithInt(item, 'id') ?? 0,
    hadithNumber: getHadithString(item, 'hadithNumber', 'hadith_number') ?? '',
    bookSlug: getHadithString(item, 'bookSlug', 'book_slug') ?? fallbackBookSlug,
    chapterNumber,
    volume: getHadithInt(item, 'volume'),
    status: getHadithString(item, 'status'),
    englishNarrator: getHadithString(item, 'englishNarrator', 'english_narrator'),
    urduNarrator: getHadithString(item, 'urduNarrator', 'urdu_narrator'),
    english: getHadithString(item, 'hadithEnglish', 'hadith_english'),
    urdu: getHadithString(item, 'hadithUrdu', 'hadith_urdu'),
    arabic: getHadithString(item, 'hadithArabic', 'hadith_arabic'),
    headingEnglish: getHadithString(item, 'headingEnglish', 'heading_english'),
    headingUrdu: getHadithString(item, 'headingUrdu', 'heading_urdu'),
    headingArabic: getHadithString(item, 'headingArabic', 'heading_arabic'),
  };
}

function ensureSlug(slug: string): void {
  if (slug.trim().length === 0 || slug.length > 80 || !SLUG_PATTERN.test(slug)) throw new ArgumentError('Invalid Hadith book slug.', 'slug');
}

/** Retry-After as delta-seconds or an HTTP-date, in milliseconds; null when absent or unparseable. */
function parseRetryAfter(header: string | null, now: Date): number | null {
  if (header === null) return null;
  const value = header.trim();
  if (/^\d+$/.test(value)) return Number(value) * 1000;
  const date = Date.parse(value);
  return Number.isFinite(date) ? date - now.getTime() : null;
}

function discardBody(response: Response): void {
  void response.body?.cancel().catch(() => undefined);
}

/**
 * Typed client for https://hadithapi.com/api/ (3 minute timeout, UA DeenTime/1.0).
 * The API key travels as the `apiKey` query parameter, so request URLs are never
 * logged. Requests are serialized with 850 ms spacing, honour 429 Retry-After and
 * retry transport failures, exactly like HadithProviderClient.cs.
 */
@Injectable()
export class HadithProviderClient {
  static readonly providerName = 'hadithapi';
  static readonly languages: readonly string[] = ['ar', 'en', 'ur'];
  private static readonly timeoutMs = 3 * 60_000;
  private static readonly maxAttempts = 8;
  private static readonly spacingMs = 850;

  private readonly fetch: typeof fetch;
  private readonly baseUrl: string;
  private readonly now: () => Date;
  private readonly delay: (ms: number, signal?: AbortSignal) => Promise<void>;
  private gate: Promise<void> = Promise.resolve();
  private nextRequestAtMs = 0;

  constructor(
    private readonly options: IslamicContentOptions,
    @Optional() @Inject(PROVIDER_CLIENT_OPTIONS) clientOptions?: ProviderClientOptions,
  ) {
    this.fetch = resolveFetch(clientOptions);
    this.baseUrl = ensureTrailingSlash(clientOptions?.baseUrl ?? options.hadithBaseUrl);
    this.now = resolveClock(clientOptions);
    this.delay = clientOptions?.delay ?? realDelay;
  }

  get isConfigured(): boolean {
    return this.options.isHadithConfigured;
  }

  async getBooks(signal?: AbortSignal): Promise<HadithProviderBook[]> {
    const document = await this.getDocument('books', null, signal);
    return findHadithArray(document, 'books')
      .map(
        (item): HadithProviderBook => ({
          id: getHadithInt(item, 'id') ?? 0,
          slug: getHadithString(item, 'bookSlug', 'book_slug', 'slug') ?? '',
          name: getHadithString(item, 'bookName', 'book_name', 'name') ?? '',
          writerName: getHadithString(item, 'writerName', 'writer_name') ?? '',
          aboutWriter: getHadithString(item, 'aboutWriter', 'about_writer'),
          writerDeath: getHadithString(item, 'writerDeath', 'writer_death'),
          hadithCount: getHadithInt(item, 'hadiths_count', 'hadithCount', 'hadith_count') ?? 0,
          chapterCount: getHadithInt(item, 'chapters_count', 'chapterCount', 'chapter_count') ?? 0,
        }),
      )
      .filter((book) => book.id > 0 && book.slug.trim().length > 0);
  }

  async getChapters(bookSlug: string, signal?: AbortSignal): Promise<HadithProviderChapter[]> {
    ensureSlug(bookSlug);
    const document = await this.getDocument(`${bookSlug}/chapters`, null, signal);
    return findHadithArray(document, 'chapters')
      .map(
        (item): HadithProviderChapter => ({
          id: getHadithInt(item, 'id') ?? 0,
          bookSlug: getHadithString(item, 'bookSlug', 'book_slug') ?? bookSlug,
          chapterNumber: getHadithInt(item, 'chapterNumber', 'chapter_number') ?? 0,
          arabic: getHadithString(item, 'chapterArabic', 'chapter_arabic'),
          english: getHadithString(item, 'chapterEnglish', 'chapter_english'),
          urdu: getHadithString(item, 'chapterUrdu', 'chapter_urdu'),
        }),
      )
      .filter((chapter) => chapter.chapterNumber > 0);
  }

  async getHadithPage(bookSlug: string, page: number, pageSize = 200, signal?: AbortSignal): Promise<HadithProviderPage> {
    ensureSlug(bookSlug);
    if (page < 1) throw ArgumentError.outOfRange('page');
    const size = Math.min(200, Math.max(1, pageSize));

    const document = await this.getDocument('hadiths', { book: bookSlug, paginate: String(size), page: String(page) }, signal);
    const envelope = findHadithProperty(document, 'hadiths');
    if (asObject(envelope) === null) throw new IslamicContentProviderError('The Hadith provider returned an unexpected response.');

    const records = findHadithArray(envelope, 'data')
      .map((item) => parseHadith(item, bookSlug))
      .filter((record) => record.id > 0 && record.hadithNumber.trim().length > 0);

    return {
      currentPage: getHadithInt(envelope, 'current_page', 'currentPage') ?? page,
      lastPage: Math.max(1, getHadithInt(envelope, 'last_page', 'lastPage') ?? page),
      perPage: getHadithInt(envelope, 'per_page', 'perPage') ?? size,
      total: getHadithInt(envelope, 'total') ?? records.length,
      records,
    };
  }

  private async getDocument(relativePath: string, parameters: Record<string, string> | null, signal?: AbortSignal): Promise<unknown> {
    if (!this.isConfigured) throw new IslamicContentProviderError('The Hadith provider key has not been configured on the server.');

    const query: Array<[string, string]> = Object.entries(parameters ?? {});
    query.push(['apiKey', this.options.hadithApiKey]);
    // Carries the server-side secret: never log or surface this URL.
    const url = this.baseUrl + relativePath.replace(/^\/+/, '') + '?' + query.map(([key, value]) => `${escapeDataString(key)}=${escapeDataString(value)}`).join('&');

    for (let attempt = 1; attempt <= HadithProviderClient.maxAttempts; attempt++) {
      const outcome = await this.withGate(() => this.attempt(url, attempt, signal));
      if (outcome !== RETRY) return outcome.document;
    }
    throw new IslamicContentProviderError('The Hadith provider is currently unavailable.');
  }

  /** SemaphoreSlim(1, 1): one upstream request at a time across all callers. */
  private async withGate<T>(work: () => Promise<T>): Promise<T> {
    const previous = this.gate;
    let release: () => void = () => undefined;
    this.gate = new Promise<void>((resolve) => {
      release = resolve;
    });
    await previous;
    try {
      return await work();
    } finally {
      release();
    }
  }

  private async attempt(url: string, attempt: number, signal?: AbortSignal): Promise<AttemptOutcome> {
    const throttleMs = this.nextRequestAtMs - this.now().getTime();
    if (throttleMs > 0) await this.delay(throttleMs, signal);

    let response: Response;
    try {
      response = await this.fetch(url, {
        headers: { 'User-Agent': 'DeenTime/1.0' },
        signal: requestSignal(HadithProviderClient.timeoutMs, signal),
      });
    } catch (error) {
      return this.transportFailure(error, attempt);
    }
    this.nextRequestAtMs = this.now().getTime() + HadithProviderClient.spacingMs;

    if (response.status === 429) {
      discardBody(response);
      const retryAfterMs = parseRetryAfter(response.headers.get('retry-after'), this.now()) ?? Math.min(60, 3 * 2 ** (attempt - 1)) * 1000;
      this.nextRequestAtMs = this.now().getTime() + Math.min(120_000, Math.max(2_000, retryAfterMs));
      if (attempt === HadithProviderClient.maxAttempts) {
        throw new IslamicContentProviderError('The Hadith provider rate limit remained active after automatic retries.');
      }
      return RETRY;
    }

    if (!response.ok) {
      discardBody(response);
      throw new IslamicContentProviderError(`The Hadith provider returned HTTP ${response.status} (${httpStatusName(response.status)}).`);
    }

    let text: string;
    try {
      text = await response.text();
    } catch (error) {
      return this.transportFailure(error, attempt);
    }
    try {
      return { document: JSON.parse(text) as unknown };
    } catch (error) {
      throw new IslamicContentProviderError('The Hadith provider returned an invalid response.', { cause: error });
    }
  }

  private transportFailure(error: unknown, attempt: number): AttemptOutcome {
    const failure = classifyFetchError(error);
    if (failure === 'timeout') throw new IslamicContentProviderError('The Hadith provider timed out.', { cause: error });
    if (failure !== 'network') throw error;
    if (attempt === HadithProviderClient.maxAttempts) {
      throw new IslamicContentProviderError('The Hadith provider is currently unavailable.', { cause: error });
    }
    this.nextRequestAtMs = this.now().getTime() + Math.min(30, attempt * 2) * 1000;
    return RETRY;
  }
}
